### ISSUE TITLE
[css-color-6] Let's finally settle contrast-color() syntax

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#7937]
@fantasai and I did a fair bit of work on css-color-6 these past couple of days, but until the WG settles on syntax, we cannot have something shippable for `contrast-color()`, not even for [the restricted version we agreed to pull into css-color-5](https://github.com/w3c/csswg-drafts/issues/7361).

For background, the previous, [confusing](https://github.com/w3c/csswg-drafts/issues/7354) syntax was like:

```css
contrast-color(wheat vs bisque, var(--accent-color), olive, sienna, color(display-p3 0 1 0), maroon to wcag(4.5))
```

The [current syntax](https://drafts.csswg.org/css-color-6/#colorcontrast) is (keywords ridiculous on purpose, as we [resolved](https://github.com/w3c/csswg-drafts/issues/7359#issuecomment-1203072807) to have them undefined):

```css
contrast-color(wheat tbd-bg wcag(4.5), bisque, var(--accent-color), olive, sienna, color(display-p3 0 1 0), maroon)
```

We should a) resolve on three, largely (but not entirely) orthogonal things:
1. Do we want to keep this syntax or change it? 
2. What should the keywords be?
3. Do we allow authors to omit the keyword to specify a default (background)?

## Problems with current syntax and alternative syntaxes

If `<target-contrast>` is just an algorithm, the current syntax is a bit hard to parse for humans as there are a lot of sequential keywords:

```css
contrast-color(wheat tbd-bg wcag2, bisque, var(--accent-color), olive, sienna, color(display-p3 0 1 0), maroon)
```

OTOH how often are colors specified in keywords? Usually [authors use hex or functional notations](https://almanac.httparchive.org/en/2021/css#fig-23), so maybe this is not a valid concern. Regardless, another syntaxe that @fantasai and I explored was:

```css
contrast-color(wheat tbd-bg wcag2 / bisque var(--accent-color) olive sienna color(display-p3 0 1 0) maroon)
```

## Foreground/background keyword options

One thing to keep in mind for this is that the base color being the background will be *far* more common.

| Foreground | Background | Pros | Cons |
|------------|------------|-------|------|
| `foreground` | `background` | Descriptive and unambiguous | Too long |
| `fg` | `bg` | Concise | [Abbreviations](https://w3ctag.github.io/design-principles/#:~:text=Abbreviations%20are%20generally%20avoided) |
| `text` | `base` | Concise without being abbreviated | foreground is not always text, `base` is too vague |
| `under` | `over` | Concise, reads naturally | "x over y, z" has entirely opposite meaning to "over x, y, z" so order needs to matter and we cannot use commas because "over x, y, z" looks like a single list |

## Should we have a default (background)?

In the majority of cases, the base color will be a background, rather than a foreground. Which brings us to the question: should the syntax allow for *neither* keyword to be specified?

The advantage of having a default is that the syntax becomes concise regardless of which keyword pair we go with, in which case we can just go for a long descriptive keyword for foreground (like `foreground`) and be done with it. Having sensible defaults is a tenet of good usability.
The problem with having a default is that authors do not learn to think about the distinction, so in practice they may not specify it even when they have a foreground.

**Comment by Loirooriol** [comment]
I like that `contrast-color(wheat tbd-bg wcag2 / bisque var(--accent-color) olive sienna color(display-p3 0 1 0) maroon)` has a clear separation between the background color and the possible foreground colors.

But `/` as a separator inside a function seems a bit weird, why not a comma as usual?

```css
contrast-color(wheat tbd-bg wcag2, bisque var(--accent-color) olive sienna color(display-p3 0 1 0) maroon)
```

**Comment by tabatkins** [comment]
> `contrast-color(wheat tbd-bg wcag2 / bisque var(--accent-color) olive sienna color(display-p3 0 1 0) maroon)`

Losing the commas means the syntax becomes completely inextensible in the future; see 'counter-reset' for an example of this being bad. I'd strongly advise against it.

> Foreground/background keyword options

I don't recall if we considered `top`/`bottom`. They're clear, fairly short, already used, and don't have the ordering concern that `over`/`under` do.

> Should we have a default (background)?

I think this should be based solely on, when using whatever algo we think is most reasonable (and default, if possible), if the fg/bg decision makes a significant difference in the output. If they're usually pretty close and never terrible when done wrong-way, then sure, make it optional, no need to require something that usually won't matter.  But if those aren't both true, then no, we should keep it required.

**Comment by LeaVerou** [comment]
> Losing the commas means the syntax becomes completely inextensible in the future; see 'counter-reset' for an example of this being bad. I'd strongly advise against it.

FWIW there are no algorithms currently that require metadata per color, though it does give me a bit of pause if we don't have the option for the future.

> > Foreground/background keyword options
> 
> I don't recall if we considered `top`/`bottom`. They're clear, fairly short, already used, and don't have the ordering concern that `over`/`under` do.

I think that's a bit confusing as it implies order in the Y axis, not the Z axis. But feel free to add them to the summary table, the more options we have the better.

> > Should we have a default (background)?
> 
> I think this should be based solely on, when using whatever algo we think is most reasonable (and default, if possible), if the fg/bg decision makes a significant difference in the output. If they're usually pretty close and never terrible when done wrong-way, then sure, make it optional, no need to require something that usually won't matter. But if those aren't both true, then no, we should keep it required.

Right now the only algorithm that even distinguishes them is APCA. @myndex can tell us (briefly please?) how significant the difference is and whether flipping them produces a terrible result or not.

**Comment by svgeesus** [comment]
> Right now the only algorithm that even distinguishes them is APCA

Distinguishing them is the correct approach. 

![image](https://user-images.githubusercontent.com/2506926/197362609-03fe35d2-b24c-4fa4-b507-cb6781f4f3a3.png)

-41 and 38 are somewhat different, yes :) and even ignoring the light-on dark and dark-on-light polarity indicator, there is certainly a difference. Not terrible, no. 

But tossing away the "which one is text" indicator would be a **bad** design decision.

**Comment by LeaVerou** [comment]
> But tossing away the "which one is text" indicator would be a **bad** design decision.

We are not talking about tossing it away, just not making it mandatory.

**Comment by svgeesus** [comment]
> just making it optional.

That would be a bad idea. Why?

**Comment by LeaVerou** [comment]
> > just making it optional.
> 
> That would be a bad idea. Why?

To optimize the common case: background. That way the length of *both* keywords stops being a concern, as they are only needed in the rare case.

**Comment by nt1m** [comment]
Personally I'm not fond of including specific algorithm names in the syntax. This should just be abstracted behind `background` vs. `text` or something similar, and the UA (or the spec) should pick the ideal algorithm for each case.

**Comment by LeaVerou** [comment]
> Personally I'm not fond of including specific algorithm names in the syntax. This should just be abstracted behind `background` vs. `text` or something similar, and the UA (or the spec) should pick the ideal algorithm for each case.

Yeah, that would be ideal, if there was a single optimal algorithm we could use. Sadly, that is not the case. However, that's what we're doing for level 5, but only for returning white and black (no candidates). There is a lot of background here, you can start from #7553 and dig in.

**Comment by ByteEater-pl** [comment]
`fore` or `front`
and `back`
(without "ground"; these are actual words, so the rule against abbreviations would be obeyed)
maybe?

**Comment by fantasai** [comment]
What about incorporating foreground vs background into the function name?

```
contrast-background(...)
contrast-foreground(...)
```

**Comment by LeaVerou** [comment]
> What about incorporating foreground vs background into the function name?
> 
> ```
> contrast-background(...)
> contrast-foreground(...)
> ```

As I replied in IRC, I think it's unclear whether `contrast-background(red)` means "I have a red background, give me a suitable foreground color" or "Get me a suitable background color, I have a red foreground".

**Comment by Myndex** [comment]
Hi Adam @argyleink 

I would like to respond to a couple things that you mentioned in the meeting, that I see listed in the IRC log. 

> _argyle: Say apca comes out and is stable,_

The current base algorithm for apca-w3 has been in public beta since Feb 2021, and that base algorithm has not changed since then, **over two years**. Thus far `0.0.98g4g` has shown to be best for a simple pair-of-colors solution for sRGB self-illuminated displays.

> _argyle: apca3 is still not producing good results yet, still a few years out_

### _YIKES!!_
**OK you have my attention.**

There have been no issues filed nor discussions in any of the repos reporting this.

I would really like to dig down and understand what problems you are seeing.

First I'd like to establish that you're basing this on the current version of the algorithm, the base algorithm is identified as ` 0.0.98g4g `, and while there are a lot of implementations out in the wild due to the popularity, many of them are unfortunately flawed, using incorrect values or implementations.

The *only* canonical tool is: **https://apcacontrast.com**

The _only_ canonical repos are:

**GitHub:** https://github.com/Myndex/apca-w3

**npm:** ` npm i apca-w3 `


### Other tools:
There are a few third-party tools that are pretty cool, but should not be used for evaluation I have a number of caveats, as they are generally not complete implementations.

Also, there is another Myndex hosted tool that is only for experimental work, at `myndex dot com/SAPC/` but it **should not be used for general evaluation** as it is used for live experiments and results may vary.

It has been brought to my attention that **Chrome** and **Edge** are still using incorrect values and incorrect code, and therefore are **not APCA** and should not be relied upon for any purpose, nor used for any form of evaluation.

And I have been trying to chase down some of the developers of some implementations that are using the wrong code in association with the wrong look up tables. 

All of these kinds of problems I hope you recognize are simply a factor of having an open and public beta period.

----
### That Said

Nevertheless, if you are finding problems of any sort with the current version, **I really need to hear about it**. Please.

I would like to talk to you about what results you found were "not good" or that are problematic, and the circumstances and characteristics of what you perceive as not usable.

If you'd like to contact me privately, my DM's are open on Twitter at https://twitter.com/MyndexResearch

You can also reach me via the email which is listed here on my profile.


Thank you for reading, I look forward to discussing this with you soon.

Andy
